connection.js

```diff
@@ -158,24 +158,26 @@ class Connection {
                 this.processMove(gamedata);
             } */
 
-            // Don't connect to old finished games.
-            if (gamedata.phase === "finished" && !(gamedata.id in this.connected_games))
-                return;
-
-            // When a game ends, we don't get a "finished" active_game.phase. Probably since the game is no
-            // longer active.(Update: We do get finished active_game events? Unclear why I added prior note.)
-            //
             if (gamedata.phase === "finished") {
-                if (config.DEBUG) conn_log(gamedata.id, "gamedata.phase === finished");
-
-                // XXX We want to disconnect right away here, but there's a game over race condition
-                //     on server side: sometimes /gamedata event with game outcome is sent after
-                //     active_game, so it's lost since there's no game to handle it anymore...
-                //     Work around it with a timeout for now.
-                if (!this.connected_games[gamedata.id].disconnect_timeout) {
-                    console.log("Starting disconnect Timeout in Connection active_game for " + gamedata.id);
-                    this.connected_games[gamedata.id].disconnect_timeout =
-                        setTimeout(() => {  this.disconnectFromGame(gamedata.id);  }, 1000);
+                if (!(gamedata.id in this.connected_games)) {
+                    // Don't connect to old finished games.
+                    return;
+                } else {
+                    // When a game ends, we don't get a "finished" active_game.phase. Probably since the game is no
+                    // longer active.(Update: We do get finished active_game events? Unclear why I added prior note.)
+                    // Note: active_game and gamedata events can arrive in either order.
+                    //
+                    if (config.DEBUG) conn_log(gamedata.id, "active_game phase === finished");
+
+                    // XXX We want to disconnect right away here, but there's a game over race condition
+                    //     on server side: sometimes /gamedata event with game outcome is sent after
+                    //     active_game, so it's lost since there's no game to handle it anymore...
+                    //     Work around it with a timeout for now.
+                    if (!this.connected_games[gamedata.id].disconnect_timeout) {
+                        if (config.DEBUG) console.log("Starting disconnect Timeout in Connection active_game for " + gamedata.id);
+                        this.connected_games[gamedata.id].disconnect_timeout =
+                            setTimeout(() => {  this.disconnectFromGame(gamedata.id);  }, 1000);
+                    }
                 }
```

Comment:
I wonder if we should return here?

Reply:
Good catch. In fact I was able to refine it further now. Take a look. :)